test/cached_test.go

```diff
@@ -0,0 +1,107 @@
+	"github.com/gadget-inc/dateilager/pkg/cached"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var pasetoKey ed25519.PublicKey
+
+func TestPopulateCache(t *testing.T) {
+	tc := util.NewTestCtx(t, auth.Admin, 1)
+	defer tc.Close()
+
+	writeProject(tc, 1, 2)
+	writeObject(tc, 1, 1, i(2), "a", "a v1")
+	aHash := writePackedFiles(tc, 1, 1, nil, "pack/a")
+	bHash := writePackedFiles(tc, 1, 1, nil, "pack/b")
+	version, err := db.CreateCache(tc.Context(), tc.Connect(), "", 100)
+	require.NoError(t, err)
+
+	c, _, close := createTestClient(tc)
+	defer close()
+
+	tmpDir := emptyTmpDir(t)
+	defer os.RemoveAll(tmpDir)
+
+	s := cached.NewServer(tc.Context(), c, testCert(), path.Join(tmpDir, "staging"), pasetoKey)
+	require.NoError(t, s.Cached.Prepare(tc.Context()), "cached.Prepare must succeed")
+
+	_, err = s.Cached.PopulateDiskCache(tc.Context(), &pb.PopulateDiskCacheRequest{Path: path.Join(tmpDir, "test")})
+	require.NoError(t, err, "Cached.PopulateDiskCache")
+
+	verifyDir(t, path.Join(tmpDir, "test"), -1, map[string]expectedFile{
+		fmt.Sprintf("objects/%v/pack/a/1", aHash): {content: "pack/a/1 v1"},
+		fmt.Sprintf("objects/%v/pack/a/2", aHash): {content: "pack/a/2 v1"},
+		fmt.Sprintf("objects/%v/pack/b/1", bHash): {content: "pack/b/1 v1"},
+		fmt.Sprintf("objects/%v/pack/b/2", bHash): {content: "pack/b/2 v1"},
+		"versions": {content: fmt.Sprintf("%v\n", version)},
+	})
+}
+
+func TestPopulateEmptyCache(t *testing.T) {
+	tc := util.NewTestCtx(t, auth.Admin, 1)
+	defer tc.Close()
+
+	writeProject(tc, 1, 2)
+	writeObject(tc, 1, 1, i(2), "a", "a v1")
+	// no packed files, so no cache
+	version, err := db.CreateCache(tc.Context(), tc.Connect(), "", 100)
+	require.NoError(t, err)
+	assert.NotEqual(t, int64(-1), version)
+
+	c, _, close := createTestClient(tc)
+	defer close()
+
+	tmpDir := emptyTmpDir(t)
+	defer os.RemoveAll(tmpDir)
+
+	s := cached.NewServer(tc.Context(), c, testCert(), path.Join(tmpDir, "staging"), pasetoKey)
+	require.NoError(t, s.Cached.Prepare(tc.Context()), "cached.Prepare must succeed")
+
+	_, err = s.Cached.PopulateDiskCache(tc.Context(), &pb.PopulateDiskCacheRequest{Path: path.Join(tmpDir, "test")})
+	require.NoError(t, err, "Cached.PopulateDiskCache")
+
+	verifyDir(t, path.Join(tmpDir, "test"), -1, map[string]expectedFile{
+		"objects/": {content: "", fileType: typeDirectory},
+	})
+}
+
+func TestPopulateCacheToPathWithNoWritePermissions(t *testing.T) {
+	tc := util.NewTestCtx(t, auth.Admin, 1)
+	defer tc.Close()
+
+	writeProject(tc, 1, 2)
+	writeObject(tc, 1, 1, i(2), "a", "a v1")
+	writePackedFiles(tc, 1, 1, nil, "pack/a")
+	writePackedFiles(tc, 1, 1, nil, "pack/b")
+	_, err := db.CreateCache(tc.Context(), tc.Connect(), "", 100)
+	require.NoError(t, err)
+
+	c, _, close := createTestClient(tc)
+	defer close()
+
+	tmpDir := emptyTmpDir(t)
+	defer os.RemoveAll(tmpDir)
+
+	s := cached.NewServer(tc.Context(), c, testCert(), path.Join(tmpDir, "staging"), pasetoKey)
```

Comment:
For testing a GRPC server we use the GRPC test tool `buffcon` to actually write over a local connection instead of booting up a real server: https://github.com/gadget-inc/dateilager/blob/main/test/shared_test.go#L371-L385

Reply:
Word, updated to do the same